demos/mqtt/mqtt_demo_connection_sharing/mqtt_demo_connection_sharing.c

```diff
@@ -0,0 +1,1601 @@
+        }
+
+        if( retriesArePending == false )
+        {
+            LogError( ( "Connection to the broker failed, all attempts exhausted." ) );
+            returnStatus = EXIT_FAILURE;
+        }
+    } while( ( opensslStatus != OPENSSL_SUCCESS ) && ( retriesArePending == true ) );
+
+    return returnStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static int getNextFreeIndexForOutgoingPublishes( uint8_t * pIndex )
+{
+    int returnStatus = EXIT_FAILURE;
+    uint8_t index = 0;
+
+    assert( outgoingPublishPackets != NULL );
+    assert( pIndex != NULL );
+
+    for( index = 0; index < MAX_OUTGOING_PUBLISHES; index++ )
+    {
+        /* A free index is marked by invalid packet id.
+         * Check if the the index has a free slot. */
+        if( outgoingPublishPackets[ index ].packetId == MQTT_PACKET_ID_INVALID )
+        {
+            returnStatus = EXIT_SUCCESS;
+            break;
+        }
+    }
+
+    /* Copy the available index into the output param. */
+    *pIndex = index;
+
+    return returnStatus;
+}
+/*-----------------------------------------------------------*/
+
+static void cleanupOutgoingPublishAt( uint8_t index )
+{
+    assert( outgoingPublishPackets != NULL );
+    assert( index < MAX_OUTGOING_PUBLISHES );
+
+    /* Clear the outgoing publish packet. */
+    ( void ) memset( &( outgoingPublishPackets[ index ] ),
+                     0x00,
+                     sizeof( outgoingPublishPackets[ index ] ) );
+}
+
+/*-----------------------------------------------------------*/
+
+static void cleanupOutgoingPublishes( void )
+{
+    assert( outgoingPublishPackets != NULL );
+
+    /* Clean up all the outgoing publish packets. */
+    ( void ) memset( outgoingPublishPackets, 0x00, sizeof( outgoingPublishPackets ) );
+}
+
+/*-----------------------------------------------------------*/
+
+static void cleanupOutgoingPublishWithPacketID( uint16_t packetId )
+{
+    uint8_t index = 0;
+
+    assert( outgoingPublishPackets != NULL );
+    assert( packetId != MQTT_PACKET_ID_INVALID );
+
+    /* Clean up all the saved outgoing publishes. */
+    for( ; index < MAX_OUTGOING_PUBLISHES; index++ )
+    {
+        if( outgoingPublishPackets[ index ].packetId == packetId )
+        {
+            cleanupOutgoingPublishAt( index );
+            LogInfo( ( "Cleaned up outgoing publish packet with packet id %u.\n\n",
+                       packetId ) );
+            break;
+        }
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+static int handlePublishResend( MQTTContext_t * pMqttContext )
+{
+    int returnStatus = EXIT_SUCCESS;
+    MQTTStatus_t mqttStatus = MQTTSuccess;
+    uint8_t index = 0U;
+
+    assert( outgoingPublishPackets != NULL );
+
+    /* Resend all the QoS1 publishes still in the array. These are the
+     * publishes that hasn't received a PUBREC. When a PUBREC is
+     * received, the publish is removed from the array. */
```

Comment:
QoS1 publishes receive PUBACKs not PUBRECs

Reply:
Fixed